hpx/util/demangle_helper.hpp

```diff
@@ -1,73 +0,0 @@
-//  Copyright (c) 2007-2012 Hartmut Kaiser
-//
-//  Distributed under the Boost Software License, Version 1.0. (See accompanying
-//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
-
-#if !defined(HPX_UTIL_DEMANGLE_HELPER_OCT_28_2011_0410PM)
-#define HPX_UTIL_DEMANGLE_HELPER_OCT_28_2011_0410PM
-
-#include <hpx/config.hpp>
-
-// disable the code specific to gcc for now as this causes problems
-// (see #811: simple_central_tuplespace_client run error)
-#if 0 // defined(__GNUC__)
```

Comment:
Am I right in thinking that the reason why the code was #ifdefed out here is because the demangle_helper uses `abi::__cxa_demangle(typeid(T).name(), 0, 0, 0)` which returns a `char *` and this char * is deleted when the helper goes out of scope - but the `char *` is held onto by the (serialization) code that  uses it.

If this is the case, is there anything we can do to make the serialization code keep the memory alive?
For debugging, the demangled version is very useful, in the `debug::print(...)` code, the `char *` is copied into a string, so no memory violations occur. Perhaps we should just use the undemangled version in the serialization, and keep the demangled one for debug stuff?

Reply:
for that, we'd just need to implement proper resource management. `std::string`, `std::shared_ptr<char>` or `std::vector<char>` are available.

Reply:
I played with this some more and something very odd is going on. The code segfaults in function registration of pointers when the debug mangler is used, but it's not caused by the pointer being invalid AFAICT. the test `any_serialization_test_exe` fails if the GNU cxxabi demangler is used and I'm baffled as to what is going wrong.

Reply:
I think I'll leave the serialization code using the standard typeid stuff and enable the demangler for debugging and printing types, it is very useful to me and worth having, but I'll just provide two types and use one for debug and the other for serialization. Unless someone has a clue as to why the cxxabi demangler causes problems